e2e-tests/src/test/sync.rs

```diff
@@ -0,0 +1,295 @@
+/// are still able to finalize new blocks.
+#[tokio::test]
+pub async fn into_multiple_groups_of_two() -> anyhow::Result<()> {
+    const NUMBER_OF_BLOCKS_TO_WAIT: u32 = 32;
+
+    let config = setup_test();
+
+    let nodes_configs = config
+        .nodes_configs()
+        .context("unable to build configuration for test nodes")?;
+    if nodes_configs.len() < 2 {
+        return Err(anyhow!("not enough nodes in configuration"));
+    }
+
+    let groups = nodes_configs.as_slice().chunks(2);
+    let groups: Vec<Vec<NodeConfig>> = groups.fold(Vec::new(), |mut result, chunk| {
+        if chunk.len() < 2 {
+            if let Some(last) = result.last_mut() {
+                last.append(&mut chunk.to_vec());
+            }
+        } else {
+            result.push(chunk.to_vec());
+        }
+        result
+    });
+
+    let disconnect_configuration = NodesConnectivityConfiguration::from(groups);
+    let reconnect_configuration = disconnect_configuration.clone().reconnect();
+
+    perform_test(
+        nodes_configs.as_slice(),
+        nodes_configs.as_slice(),
+        nodes_configs.as_slice(),
+        disconnect_configuration,
+        reconnect_configuration,
+        NUMBER_OF_BLOCKS_TO_WAIT,
+        NUMBER_OF_BLOCKS_TO_WAIT,
+    )
+    .await
+}
+
+/// Tests if nodes are able to proceed after we divide them into two ~equal size disjoint sets. More
+/// precisely, it divides nodes into two halves, awaits for both sets to create several new blocks
+/// (which shouldn't be finalized), reconnects them and checks if nodes are still able to finalize
+/// new blocks.
+#[tokio::test]
+pub async fn into_two_equal_size_groups_with_no_quorum() -> anyhow::Result<()> {
+    const NUMBER_OF_BLOCKS_TO_WAIT: u32 = 32;
+
+    let config = setup_test();
+
+    let nodes_configs = config
+        .nodes_configs()
+        .context("unable to build configuration for test nodes")?;
+
+    let (left_side, right_side) = nodes_configs.as_slice().split_at(nodes_configs.len() / 2);
+
+    let disconnect_configuration =
+        NodesConnectivityConfiguration::from(vec![left_side.to_vec(), right_side.to_vec()]);
+    let reconnect_configuration = disconnect_configuration.clone().reconnect();
+
+    perform_test(
+        nodes_configs.as_slice(),
+        nodes_configs.as_slice(),
+        nodes_configs.as_slice(),
+        disconnect_configuration,
+        reconnect_configuration,
+        NUMBER_OF_BLOCKS_TO_WAIT,
+        NUMBER_OF_BLOCKS_TO_WAIT,
+    )
+    .await
+}
+
+/// Tests if nodes are able to proceed after we divide them into two disjoint sets, where one of
+/// them consists of a quorum. More precisely, it divides nodes into two disjoint sets (one of them
+/// contains a quorum), awaits for nodes in both sets to create several new blocks (these shouldn't
+/// be finalized), reconnects all nodes and checks if nodes are still able to finalize new blocks.
```

Comment:
Wait, if one of them has a quorum shouldn't the blocks created by them actually be finalized?

Reply:
Changed.